### PR TITLE
feat(react): remove unsafe lifecycle methods from DataTable and DatePicker

### LINKED ISSUE #4902: Remove unsafe lifecycle methods from our components
We use `UNSAFE_` lifecycle methods in several components that cause errors to be displayed in our storybook and to end-users. We should remove our usage of these components in the following components:

- [ ] DatePicker

```bash
src/components/DatePicker/DatePicker.js:309:  UNSAFE_componentWillUpdate(nextProps) {
```

- [x] StructuredList

```bash
src/components/StructuredList/StructuredList.js:142:  UNSAFE_componentWillMount() {
```

- [x] ComboBox

```bash
src/components/ComboBox/ComboBox.js:202:  UNSAFE_componentWillReceiveProps(nextProps) {
```

- [ ] DataTable

```bash
src/components/DataTable/DataTable.js:143:  UNSAFE_componentWillReceiveProps(nextProps) {
```

- [x] withState

```bash
src/tools/withState.js:16:  UNSAFE_componentWillMount() {
```

### PR DESCRIPTION
**Pull request by joshblack**
Closes #4902

Updates DataTable's usage of `UNSAFE_componentWillReceiveProps` to `componentDidUpdate`. The change required some updates to the test file, namely making sure we use the last call to render and calling update for children components to be updated.

This PR also updates `DatePicker` by converting `UNSAFE_componentWillUpdate` with `componentDidUpdate`.

#### Changelog

**New**

**Changed**

- Convert usage of `UNSAFE_componentWillReceiveProps` to `componentDidUpdate`
- Convert usage of `UNSAFE_componentWillUpdate` to `componentDidUpdate`

**Removed**

#### Testing / Reviewing

- Verify storybook samples work as intended

### CODE REVIEW

**Comment by joshblack** [comment]
Note to reviewers: this was one option I was playing around with, still trying to see if it's the best route versus `getDerivedStateFromProps`

File: packages/react/src/components/DatePicker/DatePicker.js [hunk 1]
```diff
@@ -405,6 +393,19 @@ export default class DatePicker extends Component {
...
     }
+
+    // Coordinate when the given `value` prop changes. When this happens, we
+    // should update the calendar to the new value.
+    if (prevValue !== value) {
+      if (this.cal) {
+        this.cal.setDate(this.props.value);
+        this.updateClassNames(this.cal);
+      } else {
+        if (this.inputField) {
```
Comment (emyarod) [inline packages/react/src/components/DatePicker/DatePicker.js]:
```suggestion
      } else if (this.inputField) {
```

just wondering is there any reason for splitting it up like this?

**Review by emyarod — APPROVED** [review]
looks good to me

File: packages/react/src/components/DatePicker/DatePicker.js (hunk 1 above)
Comment (joshblack) [reply packages/react/src/components/DatePicker/DatePicker.js]: No reason! Just copying over from previous source. Great catch, let me add this in.

**Comment by joshblack** [comment]
@aledavila let me know if you want to sync up on reviewing or have any questions!

**Review by aledavila — APPROVED** [review]
@joshblack got it thanks for asking though ! lgtm

**Pull request merged**